File: src/SOLID_2/go/srp.go
package main

import "fmt"

// Employee represents basic Employee Structure
// has responsibility of creating Employee
type Employee struct {
	Name string
	salary float64
}

// has responsibility of calculating Employee salary
type SalaryCalculator struct {}

func (sc SalaryCalculator) Calculate(e *Employee) float64 {
	return e.salary * 1.1
}


// saving employee file to the database
type EmployeeRepository struct {}

func (er EmployeeRepository) Save(e Employee) {
	fmt.Printf("Saving %s to file \n", e.Name)
}

// Display the Employee Info
type ReportPrinter struct{}

func (rp ReportPrinter) Print(e Employee) {
	fmt.Printf("Employee: %s, Salary: %.2f \n", e.Name, e.salary)
}

// func main() {
//     employee := Employee{Name: "Natnael Yohanes", salary: 5000}
// 	calculator := SalaryCalculator{}
// 	repository := EmployeeRepository{}
// 	printer := ReportPrinter{}

// 	employee.salary = calculator.Calculate(&employee)
// 	repository.Save(employee)
// 	printer.Print(employee)
// }

File: src/SOLID_2/ts/SRP.ts
// Violating SRP
//! Bad Implementation
// class EmployeeManager {
//     fun calculateSalary() { /* ... */ }
//     fun saveToFile() { /* ... */ }
//     fun printReport() { /* ... */ }
// }

// the class does too much — 3 responsibilities:

// Business logic (salary)

// Persistence (file saving)

// Presentation (report printing)

// only for creating an EMployee
class Employee {
  constructor(public name: string, public salary: number) {}
}

// For calculating the salary of the Employee
class SalaryCalculator {
  calculate(employee: Employee): number {
    return employee.salary * 1.1;
  }
}

// for saving mostly related to database
class EmployeeRepository {
  save(employee: Employee): void {
    console.log(`Saving ${employee.name} to file...`);
  }
}

// Display the Employee Information
class ReportPrinter {
  print(employee: Employee): void {
    console.log(`Employee ${employee.name}, Salary: ${employee.salary}`);
  }
}

const employee = new Employee("Natnael Yo", 5000);
const calculator = new SalaryCalculator();
const repository = new EmployeeRepository();
const printer = new ReportPrinter();

employee.salary = calculator.calculate(employee);
repository.save(employee);
printer.print(employee);
